edb/server/mng_port/edgecon.pyx

```diff
@@ -1447,3 +1471,293 @@ cdef class EdgeConnection:
+
+                msg_buf.write_len_prefixed_bytes(block.type_desc)
+
+                desc = blocks_desc[block.schema_object_id]
+                msg_buf.write_int64(desc['count'])
+                msg_buf.write_int64(desc['size'])
+
+            msg_buf.end_message()
+            self._transport.write(msg_buf)
+
+        finally:
+            for pgcon in pgcons:
+                pgcon.terminate()
+
+    async def restore(self):
+        cdef:
+            WriteBuffer msg_buf
+            char mtype
+
+        if self.dbview.txid:
+            raise errors.ProtocolError(
+                'RESTORE must not be executed while in transaction'
+            )
+
+        self.reject_headers()
+        schema_ddl = self.buffer.read_len_prefixed_bytes()
+        self.buffer.read_int16()  # discard -j level
+        block_num = <uint32_t>self.buffer.read_int32()
+        blocks = []
+        for _ in range(block_num):
+            blocks.append((
+                self.buffer.read_bytes(16),
+                self.buffer.read_len_prefixed_bytes(),
+            ))
+
+        self.buffer.finish_message()
+        dbname = self.dbview.dbname
+        pgcon = await self.port.new_pgcon(dbname)
+
+        try:
+            await pgcon.simple_query(
+                b'''START TRANSACTION
+                        ISOLATION LEVEL SERIALIZABLE;
+                ''',
+                True
+            )
+            tx_snapshot_id = await pgcon.simple_query(
+                b'SELECT pg_export_snapshot();', False)
+            tx_snapshot_id = tx_snapshot_id[0][0].decode()
+
+            schema_sql_units, restore_blocks, tables = \
+                await self.get_backend().compiler.call(
+                    'describe_database_restore',
+                    tx_snapshot_id,
+                    schema_ddl,
+                    blocks,
+                )
+
+            for query_unit in schema_sql_units:
+                if query_unit.system_config:
+                    raise RuntimeError(
+                        'system config commands are not supported '
+                        'during restore')
+                elif query_unit.config_ops:
+                    raise RuntimeError(
+                        'config commands are not supported '
+                        'during restore')
+                else:
+                    await pgcon.simple_query(
+                        b';'.join(query_unit.sql),
+                        ignore_data=True)
+
+            restore_blocks = {
+                b.schema_object_id: b.sql_copy_stmt
+                for b in restore_blocks
+            }
+
+            disable_trigger_q = ''
+            enable_trigger_q = ''
+            for table in tables:
+                disable_trigger_q += (
+                    f'ALTER TABLE {table} DISABLE TRIGGER ALL;'
+                )
+                enable_trigger_q += (
+                    f'ALTER TABLE {table} ENABLE TRIGGER ALL;'
```

Comment:
This might conflict with a hack in constraint trigger logic.  It shouldn't cause an error, but would make things slower, as there would be an unnecessary trigger invocation.  I will see if it's possible to avoid using `DISABLE TRIGGER` on my end.

Reply:
>  I will see if it's possible to avoid using DISABLE TRIGGER on my end.

I can just drop this for now. Disabling/enabling triggers didn't have any effect on performance of our movies db dump/restore cycle.  Is it important to disable triggers?

Reply:
Triggers might misfire depending on the order in which we restore stuff, so it's best to disable them for restore.